src/controllers/untitledSqlDocumentService.ts

```diff
@@ -9,22 +9,78 @@ import VscodeWrapper from "./vscodeWrapper";
 /**
  * Service for creating untitled documents for SQL query
  */
-export default class UntitledSqlDocumentService {
-    constructor(private vscodeWrapper: VscodeWrapper) {}
+export default class UntitledSqlDocumentService implements vscode.Disposable {
+    public trackedUris: Set<string> = new Set();
+    private _pendingOperations: Map<string, Promise<vscode.TextEditor>> = new Map();
+    private _documentCloseDisposable: vscode.Disposable | undefined;
+
+    constructor(private vscodeWrapper: VscodeWrapper) {
+        this.setupDocumentCloseListener();
+    }
+
+    /**
+     * Set up listener for document close events to automatically untrack URIs
+     */
+    private setupDocumentCloseListener(): void {
+        this._documentCloseDisposable = vscode.workspace.onDidCloseTextDocument((document) => {
+            const uri = document.uri.toString();
+            if (this.trackedUris.has(uri)) {
+                this.trackedUris.delete(uri);
+            }
+        });
+    }
+
+    dispose() {
+        if (this._documentCloseDisposable) {
+            this._documentCloseDisposable.dispose();
+        }
+    }
+
+    /**
+     * Wait for all pending operations to complete
+     */
+    public async waitForAllOperations(): Promise<vscode.TextEditor[]> {
+        const pendingPromises = Array.from(this._pendingOperations.values());
+        return Promise.all(pendingPromises);
+    }
 
     /**
      * Creates new untitled document for SQL query and opens in new editor tab
      * with optional content
      */
-    public async newQuery(content?: string): Promise<vscode.TextEditor> {
-        // Open an untitled document. So the  file doesn't have to exist in disk
-        let doc = await this.vscodeWrapper.openMsSqlTextDocument(content);
-        // Show the new untitled document in the editor's first tab and change the focus to it.
+    public async newQuery(content?: string, trackUri: boolean = true): Promise<vscode.TextEditor> {
+        // Create a unique key for this operation to handle potential duplicates
+        const operationKey = `${Date.now()}-${Math.random()}`;
+
+        try {
+            const operationPromise = this.createDocument(content, trackUri);
+            this._pendingOperations.set(operationKey, operationPromise);
```

Comment:
Is there any risk of `onDidOpenTextDocument` (and therefore `waitForAllOperations()` getting executed before this is added as a pending operation?  Maybe `createDocument()` should be sync instead of async and should  have its logic wrapped in a Promise that's returned immediately.